### PR TITLE
Feature assign driver to buses

### PR DESCRIPTION
**Pull request by fab-ryan**
### What does PR do?
As **an operator**, I want to assign a driver to a bus, so that they can control it.

### Description of Task Completed

`GET /api/v1/assign/`

`POST /api/v1/assign/bus/:busId/driver/:driverId`

`GET /api/v1/assign/:assignId`

`GET /api/v1/assign?page=number&size=number`

`PATCH /api/v1/assign/:assignId`

`DELETE /api/v1/assign/:assignId`

### How should this be manually test

When team needs to Integrate it with frontend this how run test `npm run test` 

- Using Postman test every endpoint above this header
key: Content-Type value: `application/json`

     Authorization value: JWT `Bearer token`

     Accept-Language value: `en,kiny,kisw,fr`

### Background context finished in this API
- [x] The operator should be able to assign a driver to a bus.
- [x] The operator should be able to view a paginated list of driver to bus assignments
- [x] The driver should receive a bus assignment and un assignment email & in-app notifications.
- [x] Appropriate error messages are displayed on unsuccessful attempts

### CODE REVIEW

**Review by aimelive — APPROVED** [review]
Reviewed!

File: __test__/assign.test.js [hunk 1]
```diff
@@ -0,0 +1,127 @@
+import chai from 'chai';
+import chaiHttp from 'chai-http';
+import server from '../src/server';
+import model from '../src/database/models/index';
+chai.use(chaiHttp);
+chai.should();
```
Comment (Anguandia) [inline __test__/assign.test.js]: Test setup could be centralized, instead of repeated per test file

File: __test__/assign.test.js [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+
+const token = `Bearer ${process.env.admintoken}`;
```
Comment (Anguandia) [inline __test__/assign.test.js:3]: How are you setting a dynamically generated token in the environment?

File: __test__/assign.test.js [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+      .request(server)
+      .post('/api/v1/users/login')
+      .send({
+        email: 'operator@test.com',
+        password: 'pass12345',
+      })
+      .end((error, response) => {
+        chai.expect(response.statusCode).to.equal(200);
+        OperatorToken = `Bearer ${response.body.token}`;
+        chai.expect(response.body).to.have.property('token');
```
Comment (Anguandia) [inline __test__/assign.test.js]: Please only expect in test cases (it blocks), not in hooks

File: __test__/assign.test.js [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+      .send(data)
+      .then((res) => {
+        chai.expect(res).to.have.status(201);
+      });
+    done();
+  });
+  it('Get One Driver Assigned to Buse with status code 500', (done) => {
+    chai
+      .request(server)
+      .get('/api/v1/assign/a')
```
Comment (Anguandia) [inline __test__/assign.test.js:94]: Not sure if this simulates an internal error, would be interesting to see that though. At your free time, lookup mocks, spies, etc

File: __test__/auth.test.js [hunk 5]
```diff
@@ -47,13 +47,12 @@ describe('TESTING USER AUTHENTICATION', () => {
         password: '',
       })
       .end((error, response) => {
-        chai.expect(response.statusCode).to.equal(400);
+        // chai.expect(response.statusCode).to.equal(400);
```
Comment (Anguandia) [inline __test__/auth.test.js:50]: Why remove this verification, perhaps the code could only be more specific

File: __test__/bus.test.js [hunk 6]
```diff
@@ -1,50 +1,52 @@
-import model from '../src/database/models'
+import model from '../src/database/models';
 import Op from 'sequelize';
-import chai from 'chai'
-import chaiHttp from 'chai-http'
-import server from '../src/server'
+import chai from 'chai';
+import chaiHttp from 'chai-http';
+import server from '../src/server';
 const token = `Bearer ${process.env.ADMIN_TOKEN}`;
```
Comment (Anguandia) [inline __test__/bus.test.js]: When did you set this token to the environment?

File: src/controllers/authController.js [hunk 7]
```diff
@@ -5,13 +5,13 @@ import models from '../database/models';
 import { promisify } from 'util';
 import catchAsync from '../utils/catchAsync';
 
-const signToken = (id) =>
-  jwt.sign({ id }, process.env.JWT_SECRET, {
+const signToken = (user) =>
+  jwt.sign({ user }, process.env.JWT_SECRET, {
```
Comment (Anguandia) [inline src/controllers/authController.js:9]: Can't seem to see where dotenv is set up. Please check if environment variables are actually being read, not only here, in tests, and other related files. Consider refactoring .env set up to an application-wide file

File: src/controllers/authController.js [hunk 8]
```diff
@@ -70,16 +69,26 @@ exports.protect = catchAsync(async (req, res, next) => {
...
+  //1 to check if user is OPerator Only
+  token = req.headers.authorization.split(' ')[1];
+  const decoded = await promisify(jwt.verify)(token, process.env.JWT_SECRET);
+  if (decoded.user.role !== 'operator') {
+    return next(
+      new AppError("You don't have permission to perform this task ", 401)
+    );
+  }
+  next();
+};
```
Comment (Anguandia) [inline src/controllers/authController.js]: These are typically middleware functions, authController is a misnomer, this should be a middleware. Please discuss with the team and agree

File: src/database/seeders/20220418183506-demo-user.js [hunk 9]
```diff
@@ -25,6 +25,16 @@ module.exports = {
...
+          id: 9,
+          firstName: 'Fabrice',
+          lastName: 'NDACYAYISENGA',
+          email: 'operator@test.com',
+          role: 'operator',
+          password: await bcrypt.hash('pass12345', 12),
+          createdAt: new Date(),
+          updatedAt: new Date(),
+        },
       ],
```
Comment (Anguandia) [inline src/database/seeders/20220418183506-demo-user.js:38]: Please consider creating an npm script for running seeders

File: src/routes/api/assignDriverRoutes.js [hunk 10]
```diff
@@ -0,0 +1,42 @@
+import express from 'express';
```
Comment (Anguandia) [inline src/routes/api/assignDriverRoutes.js:1]: The file name seems somewhat misleading or confusing, what's your opinion? Drivers are assigned to buses, and buses to routes, not drivers to routes

File: src/routes/index.js [hunk 11]
```diff
@@ -2,10 +2,14 @@ import express from 'express';
...
-
+import assignDriver from './api/assignDriverRoutes';
+import unassigned from './api/unassignedRoute';
 const routes = express.Router();
+
 routes.use('/users', userRoutes);
 routes.use('/buses', crudBusRoutes);
 routes.use('/routes', busRoutes);
+routes.use('/assign', assignDriver);
+routes.use('/unassigned', unassigned);
```
Comment (Anguandia) [inline src/routes/index.js:12]: Good work. Alternatively, assign and un-assign could all be bus routes with an endpoint like this:
`api/v1/buses/:busId/drivers/<:driverId>` where the method is still PATCH and with the driver Id is assign, and without, un-assign. This is just an alternative suggestion, if it interests you, we can discuss, but you may as well continue as is

File: src/utils/paginationHandler.js [hunk 12]
```diff
@@ -7,10 +7,10 @@ const getPagination = (page, size) => {
 
 
 const getPagingData = (data, page, limit) => {
-    const { count: totalItems, rows: stack } = data;
+    const { count: totalItems, rows: result } = data;
```
Comment (Anguandia) [inline src/utils/paginationHandler.js:10]: Excellent! you added page limit to the pagination, please keep it up

**Review by Anguandia — CHANGES_REQUESTED** [review]
Good work, a lot of dedication and a great amount of work done. Apart from the specific in-code comments and change requests, it will be good to consider the following general comments

- Please improve the PR description and harmonize it with team/internal conventions/best practices [here-in](https://github.com/atlp-rwanda/engineering-playbook/tree/master/5.%20Developing/Conventions#branch-naming)
- Test manually locally (with postman) to ensure no feature is broken. Register user is not working
- Tests are not running, please investigate everywhere environment variables are read, in both tests and code. Hint: look into dotenv setup
- You need to clear any junk (data) created by tests after running the tests, else subsequent runs may be greatly impacted by uncleaned residual test data generated by previous runs. Please consider the use of after hooks for this purpose
- Please ensure test quality, it is not sufficient to just test the status code of a response, but also the value of a particular property in the response
- Please do not forget squashing your commits
- Another PR has been merged, please remember to rebase

File: __test__/assign.test.js (hunk 2 above)
Comment (fab-ryan) [reply __test__/assign.test.js:3]: I used this for testing incorrect authentication and incorrect role

**Review by GilbertUp — APPROVED** [review]

File: src/routes/index.js (hunk 11 above)
Comment (fab-ryan) [reply src/routes/index.js:12]: Hello What if I make it like this `api/v1/assign/buses/:busid/drivers/:/driverId`

File: src/utils/paginationHandler.js (hunk 12 above)
Comment (fab-ryan) [reply src/utils/paginationHandler.js:10]: The page Limits are getPagination function and after it sort data result those results in getPagingData 
```suggestion
    const getPagination = (page, size) => {
    const limit = size ? + size : 3;
    const offset = page ? page * limit : 0;
    return { limit, offset };
};
```

File: src/database/seeders/20220418183506-demo-user.js (hunk 9 above)
Comment (fab-ryan) [reply src/database/seeders/20220418183506-demo-user.js:38]: I did that but I have commit it yet

File: src/routes/index.js (hunk 11 above)
Comment (fab-ryan) [reply src/routes/index.js:12]: I have finished this request

File: __test__/bus.test.js (hunk 6 above)
Comment (fab-ryan) [reply __test__/bus.test.js]: I have set them in .env

File: src/controllers/authController.js (hunk 7 above)
Comment (fab-ryan) [reply src/controllers/authController.js:9]: I didn't do this task 
about user Login and register

File: src/routes/api/assignDriverRoutes.js (hunk 10 above)
Comment (fab-ryan) [reply src/routes/api/assignDriverRoutes.js:1]: No Name it like that because it is router sections

File: package.json [hunk 13]
```diff
@@ -27,7 +27,8 @@
       }
     },
     "test": {
-      "command": "nyc mocha --require @babel/register ./__test__/ --recursive --timeout 20000 --exit",
+      "command":"sequelize-cli db:migrate:undo:all ; sequelize-cli db:migrate ; sequelize-cli db:seed:all ; nyc mocha --require @babel/register ./__test__/ --recursive --timeout 20000 --exit",
+      
       "env": {
         "NODE_ENV": "test",
         "PORT": 5000
```
Comment (Anguandia) [inline package.json]: It too, would be good to get these variables (port) from the environment, lest you'll cause conflicts in environments other than your local

**Comment by Anguandia** [comment]
I don't think statically assigning ports in the better script is a good idea, just remove the ports so they can be read from the environment

**Review by Anguandia — APPROVED** [review]

**Review by furebo — APPROVED** [review]

**Pull request merged**